File: 1. Study/0606 - 자료구조(스택)/kimshin/1. 올바른 괄호.py
def sol(s):
    answer = 'YES'
    stack=[]

    for x in s:
        if x == '(':
            stack.append(x)
        else:
            if len(stack) ==0:
                return 'NO'
            stack.pop()
    if len(stack) > 0:
        answer = 'NO'
    return answer

a = "(()(()))(()"
print(sol(a))

# https://velog.io/@soceu/%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98section-6-2%EA%B4%84%ED%98%B8%EB%AC%B8%EC%9E%90%EC%A0%9C%EA%B1%B0%EC%8A%A4%ED%83%9D

File: 1. Study/0606 - 자료구조(스택)/mojaeya/1_올바른괄호.js
console.log(solution("(()(()))(()"));

function solution(s) {
  let stack = [];
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") {
      stack.push("(");
    } else {
      stack.pop();
    }
  }
  if (stack.length > 0) return "NO";
  else return "YES";
}
